File: lib/rgfa/lines.rb
require_relative "error"

RGFA::Lines = Module.new

require_relative "lines/headers"
require_relative "lines/collections"
require_relative "lines/creators"
require_relative "lines/destructors"
require_relative "lines/finders"

#
# Methods for the RGFA class, which allow to handle lines of multiple types.
#
module RGFA::Lines

  include RGFA::Lines::Headers
  include RGFA::Lines::Collections
  include RGFA::Lines::Creators
  include RGFA::Lines::Destructors
  include RGFA::Lines::Finders

  GFA1Specific = [
                   RGFA::Line::Edge::Link,
                   RGFA::Line::Edge::Containment,
                   RGFA::Line::Group::Path,
                   RGFA::Line::Segment::GFA1,
                 ]

  GFA2Specific = [
                   RGFA::Line::CustomRecord,
                   RGFA::Line::Fragment,
                   RGFA::Line::Gap,
                   RGFA::Line::Edge::GFA2,
                   RGFA::Line::Segment::GFA2,
                   RGFA::Line::Group::Unordered,
                   RGFA::Line::Group::Ordered,
                   RGFA::Line::Unknown,
                  ]

  private

  def api_private_check_gfa_line(gfa_line, callermeth)
    if !gfa_line.kind_of?(RGFA::Line)
      raise RGFA::TypeError,
        "Note: ##{callermeth} is API private, do not call it directly\n"+
        "Error: line class is #{gfa_line.class} and not RGFA::Line"
    elsif gfa_line.rgfa != self
      raise RGFA::RuntimeError,
        "Note: ##{callermeth} is API private, do not call it directly\n"+
        "Error: line.rgfa is "+
        "#{gfa_line.rgfa.class}:#{gfa_line.rgfa.object_id} and not "+
        "RGFA:#{self.object_id}"
    end
  end

end

File: gfapy/lines/lines.py
import gfapy
import gfapy.line
from .collections import Collections
from .headers import Headers
from .creators import Creators
from .destructors import Destructors
from .finders import Finders

class Lines(Collections, Creators, Destructors, Finders, Headers):

  GFA1Specific = [
                   gfapy.line.edge.Link,
                   gfapy.line.edge.Containment,
                   gfapy.line.group.Path,
                   gfapy.line.segment.GFA1
                 ]

  GFA2Specific = [
                   gfapy.line.CustomRecord,
                   gfapy.line.Fragment,
                   gfapy.line.Gap,
                   gfapy.line.edge.GFA2,
                   gfapy.line.segment.GFA2,
                   gfapy.line.group.Unordered,
                   gfapy.line.group.Ordered,
                   gfapy.line.Unknown
                 ]

  def _api_private_check_gfa_line(self, gfa_line, callermeth):
    if not isinstance(gfa_line, gfapy.Line):
      raise gfapy.TypeError("Note: {} is API private, ".format(callermeth)+
          "do not call it directly\n"+
          "Error: line class is {} and not gfapy.Line")
    elif not gfa_line._gfa is self:
      raise gfapy.RuntimeError("Note: {} is API private, ".format(callermeth)+
          "do not call it directly\n"+
          "Error: line.gfa is not the expected instance of gfapy.Gfa\n"+
          repr(gfa_line.gfa)+" != "+repr(self))
